R/Spatial.R

```diff
@@ -0,0 +1,180 @@
+#' @import Seurat ggplot2 Matrix
+#' @importFrom compositions geometricmean geometricmeanRow 
+#' @importFrom RUVSeq RUVg
+#' @importFrom matrixStats rowMedians
+
+#' @title Q3_Normalization
+#'
+#' @description a normalization method that uses the 75th percentile to find size factors for data normalization.
+#' @param seuratObj The Seurat object that holds the data to be normalized.
+#' @param assay The assay holding the raw data - the counts slot will be assumed to hold raw data.
+#' @return Returns a Seurat object containing normalized counts stored in the counts slot of the Q3 assay. 
+#' @export
+
+Q3_Normalization <- function(seuratObj, assay = "RNA"){
+  #from here: https://www.dlongwood.com/wp-content/uploads/2021/05/WP_MK2833_Introduction_to_GeoMx_Normalization_CTA_R5-1.pdf page 7
+  # first, divide all the genes per AOI by their respective Q3 count, 
+  #ii) second, multiply all the genes in all AOIs by a constant, defined as the geometric mean of Q3 counts for all AOIs.
+  
+  #calculate column (ROI) Q3s
+  counts <- as.matrix(seuratObj[[assay]]@counts)
+  Q3_Sample_Counts <- apply(counts,MARGIN = 2,FUN = quantile,prob=c(.75))
+  #Ensure no downstream division by zero errors
+  if(any(Q3_Sample_Counts == 0)){
+    stop("Error: 75th percentile is zero for some samples. Q3 is either too low to normalize by, or the data is too sparse.")
+  }
+  #divide count matrix by vector of Q3s
+  Q3_Divided_Counts <- counts/Q3_Sample_Counts
+  #multiply Q3 divided counts by geometric mean of Q3s (scalar)
+  Q3_Final_Counts <- Q3_Divided_Counts * compositions::geometricmean(Q3_Sample_Counts)
+  
+  #stash into seurat object under new assay
+  seuratObj[['Q3']] <- Seurat::CreateAssayObject(as(Q3_Final_Counts, "dgCMatrix"))
```

Comment:
Two things: 1) if you're doing this you might as well make the targetAssayName an argument. you can make it's default value 'Q3', so usually people can omit it. However, why not store the results in @data of the original assay? Isnt that more normal?

Reply:
agreed on targetAssayName, I'll change that. 

On storing normalization in @data, it is definitely more normal, but I think it's a good trap to fall into before I've coalesced into a "default" normalization strategy like we have with scRNASeq & ADTs. This way it's much harder to confuse normalization results since they're stored in different assays.

Reply:
I think I can add more slots than just counts, data, scale.data by creating a new class (discussion: https://github.com/satijalab/seurat/discussions/5617, example: https://github.com/stuart-lab/signac/blob/63d61bed2d6f43c7fd483b77eef0c84aa70952c9/R/objects.R#L98-L111), but I'm not sure if there is anything downstream that checks for class name that might trip this up.

Reply:
I don't work with this much in R, but perhaps you can make a subclass of assay, which adds other slots? That's the normal pattern you'd follow in most programming languages.

Reply:
I think we agreed on storing the data in assays until there's a "normalization of record" to put into the @data slot. Is that right? if so, I'll resolve this.

Reply:
yes